src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -32,38 +33,41 @@ public static class ActivatorUtilities
             params object[] parameters)
         {
             int bestLength = -1;
-            bool seenPreferred = false;
-
             ConstructorMatcher bestMatcher = default;
 
             if (!instanceType.IsAbstract)
             {
-                foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
+                ConstructorInfo[] ctors = instanceType.GetConstructors();
+                ConstructorInfo[] preferredCtors = ctors
+                    .Where(x => x.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute))).ToArray();
+                if (preferredCtors.Length > 1)
                 {
-                    var matcher = new ConstructorMatcher(constructor);
-                    bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
+                    ThrowMultipleCtorsMarkedWithAttributeException();
+                }
 
-                    if (isPreferred)
+                if (preferredCtors.Length == 1)
+                {
+                    bestMatcher = new ConstructorMatcher(preferredCtors[0]);
```

Comment:
It is wasteful to construct an array, then only use its first element and throw away the rest. This block can be structured like this:
```c#
ConstructorInfo? constructorInfo = null;
foreach (ConstructorInfo? ctor in ctors)
{
    if (ctor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute)))
    {
        if (constructorInfo is not null)
        {
            ThrowMultipleCtorsMarkedWithAttributeException();
        }
        constructorInfo = ctor;
    }
}

if (constructorInfo is not null)
{
    ConstructorMatcher bestMatcher = new(constructorInfo);
    bestLength = bestMatcher.Match(parameters);
    if (bestLength == -1)
    {
        ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
    }
}
```
This will also make `using System.Linq;` unnecessary.

Reply:
@am11 Thanks for review. Good point